apps/src/applab/firebaseCounters.js

```diff
@@ -0,0 +1,125 @@
+ * Each `interval` is a number of seconds during which at most `RATE_LIMIT[interval]`
+ * writes can be performed on the data in the current channel. We store counters
+ * `writeCount` and `lastResetTime` in firebase for each channel and each interval.
+ *
+ * `writeCount` is incremented prior to every table write and can never
+ * exceed RATE_LIMIT[interval]. writeCount can be reset to 0 (or 1), but
+ * only if at least `interval` seconds have passed since the last reset.
+ *
+ * `lastResetTime` keeps track of when the last reset happened.
+ *
+ * @returns {Promise<>} Promise which succeeds if the rate limit counters are
+ *   successfully updated, or which fails with an error message if one of the
+ *   limits is exceeded.
+ */
+export function incrementRateLimitCounters() {
+  return getCurrentTime().then(currentTimeMs => {
+    let promises = [];
+    Object.keys(RATE_LIMITS).forEach(interval => {
+      const limitRef = getDatabase(Applab.channelId).child(`limits/${interval}`);
+      promises.push(limitRef.transaction(limitData => {
+        limitData = limitData || {};
+        limitData.lastResetTime = limitData.lastResetTime || 0;
+        limitData.writeCount = (limitData.writeCount || 0) + 1;
+        if (limitData.writeCount <= RATE_LIMITS[interval]) {
+          return limitData;
+        } else if (limitData.lastResetTime + interval * 1000 < currentTimeMs) {
+          // The maximum number of writes has been exceeded in more than `interval` seconds.
+          // Reset the counters.
+          limitData.writeCount = 1;
+          limitData.lastResetTime = Firebase.ServerValue.TIMESTAMP;
+          return limitData;
+        } else {
+          // The maximum number of writes has been exceeded in less than `interval` seconds.
+          // Abort the transaction.
+          return;
+        }
+      }).then(transactionData => {
+        if (!transactionData.committed) {
+          const lastResetTimeMs = transactionData.snapshot.child('lastResetTime').val();
+          const nextResetTimeMs = lastResetTimeMs + interval * 1000;
+          const timeRemaining = Math.ceil((nextResetTimeMs - currentTimeMs) / 1000);
+          // TODO(dave): notify new relic
+          return Promise.reject(`rate limit exceeded. please wait ${timeRemaining} seconds before retrying.`);
+        } else {
+          return Promise.resolve();
+        }
+      }));
+    });
+    return Promise.all(promises);
+  });
+}
+
+/**
+ * Obtains the server time by writing the current time to a node and then reading it.
+ * This node is removed when the user disconnects.
+ * @returns {Promise<number>} The current server time in milliseconds.
+ */
+function getCurrentTime() {
+  const serverTimeRef = getDatabase(Applab.channelId)
+    .child(`server_time/${Applab.firebaseUserId}`);
+  return serverTimeRef.set(Firebase.ServerValue.TIMESTAMP).then(() => {
+    serverTimeRef.onDisconnect().remove();
+    return serverTimeRef.once('value').then(snapshot => snapshot.val());
```

Comment:
It will, but we are only using this timestamp to make sure that at least 15 (or 60, or a larger number) of seconds have passed, so issues where latency or another competing write make you off by a few hundred ms shouldn't impact behavior negatively. 

I borrowed the approach from here: http://jsfiddle.net/firebase/VBmA5/